src/utils.js

```diff
@@ -209,3 +216,100 @@ export function createModelledFunctionCall(
     }
   };
 }
+
+export function isBindingMutationOutsideFunction(
+  binding: Binding,
+  bindingEntry: BindingEntry,
+  effects: Effects,
+  F: FunctionValue
+): boolean {
+  // If either the previous values are uninitialized then this is not a mutation
+  if (binding.value === undefined || bindingEntry.value === undefined) {
+    return false;
+  }
+  // If the binding has a "strict" property then it's const and can never mutate
+  if (binding.strict !== undefined) {
+    return false;
+  }
+  let env = F.$Environment;
+  let bindingName = binding.name;
+  let bindingEnv = binding.environment;
+  // If the bindingEnv is the same env, we know the mutation was only local
+  if (bindingEnv instanceof FunctionEnvironmentRecord && bindingEnv.$FunctionObject === F) {
+    return false;
+  }
+  // Check if the binding was mutated in a scope of F or its parents
+  if (env instanceof LexicalEnvironment) {
+    while (env !== null) {
+      let envRecord = env.environmentRecord;
+
+      if (envRecord instanceof GlobalEnvironmentRecord) {
+        for (let name of envRecord.$VarNames) {
+          if (name === bindingName) {
+            return true;
+          }
+        }
+        envRecord = envRecord.$DeclarativeRecord;
+      }
+      if (envRecord instanceof DeclarativeEnvironmentRecord) {
+        if (envRecord.bindings[bindingName] === binding) {
+          return true;
+        }
+      }
+      env = env.parent;
+    }
+  }
+  let functionBindingWasCreatedIn = bindingEnv.creatingOptimizedFunction;
+  // If it was never created in an optimized function then it was created in a scope
+  // of F and was mutated outside of it.
+  if (functionBindingWasCreatedIn === undefined) {
+    return true;
+  }
+  env = functionBindingWasCreatedIn.$Environment;
+  // Check if the function the binding was created in is a child scope of F,
+  // if it is not, then this binding was mutated outside of F.
+  while (env !== null) {
+    if (env === F.$Environment) {
+      return false;
+    }
+    env = env.parent;
+  }
+  return true;
+}
+
+export function areEffectsPure(realm: Realm, effects: Effects, F: FunctionValue): boolean {
+  for (let [binding] of effects.modifiedProperties) {
+    let obj = binding.object;
+    if (!effects.createdObjects.has(obj)) {
+      return false;
+    }
+  }
+
+  for (let [binding, bindingEntry] of effects.modifiedBindings) {
+    if (isBindingMutationOutsideFunction(binding, bindingEntry, effects, F)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+export function reportSideEffectsFromEffects(
+  realm: Realm,
+  effects: Effects,
+  F: FunctionValue,
+  sideEffectCallback: SideEffectCallback
+): void {
+  for (let [binding] of effects.modifiedProperties) {
+    let obj = binding.object;
+    if (!effects.createdObjects.has(obj)) {
+      sideEffectCallback("MODIFIED_PROPERTY", binding, binding.object.expressionLocation);
```

Comment:
This is a tricky one and one of the downsides of the approach in this PR. We'd probably have to store the location at the time of modification. Maybe that can be a follow up, as it seems that we need to take some time to consider the different approaches.

Reply:
I guess it's fine to land this without a story for location because it only really affects React Compiler stuff, but this will make it very difficult to debug any impurity issues.